firefox-ios/Client/Frontend/TrackingProtection/CertificatesHandler.swift

```diff
@@ -20,17 +19,67 @@ class CertificatesHandler {
     }
 
     /// Extracts and handles the certificate chain.
-    func handleCertificates() {
+    /// - Parameters:
+    ///   - completion: A completion block that provides the extracted certificates.
+    func handleCertificates(completion: @escaping ([Certificate]) -> Void) {
+        var certificates = [Certificate]()
         guard let certificateChain = SecTrustCopyCertificateChain(serverTrust) as? [SecCertificate] else {
+            completion(certificates)
             return
         }
-        for (_, certificate) in certificateChain.enumerated() {
+        for certificate in certificateChain {
             let certificateData = SecCertificateCopyData(certificate) as Data
             do {
                 let certificate = try Certificate(derEncoded: Array(certificateData))
                 certificates.append(certificate)
             } catch {
+                print("Error decoding certificate: \(error)")
+            }
+        }
+        completion(certificates)
+    }
+}
+
+// Define a function to get the certificates for a given URL
+func getCertificates(for url: URL, completion: @escaping ([Certificate]?) -> Void) {
+    // Create a URL session with a custom delegate
+    let session = URLSession(configuration: .default,
+                             delegate: CertificateDelegate(completion: completion),
+                             delegateQueue: nil)
+
+    // Start a data task to trigger the certificate retrieval
+    let task = session.dataTask(with: url) { _, _, error in
+        if let error = error {
+            print("Error fetching data: \(error)")
+            completion(nil)
+        }
+    }
+
+    task.resume()
+}
+
+// Custom delegate to handle the authentication challenge
+class CertificateDelegate: NSObject, URLSessionDelegate {
+    private let completion: ([Certificate]?) -> Void
+
+    init(completion: @escaping ([Certificate]?) -> Void) {
+        self.completion = completion
+    }
+
+    func urlSession(_ session: URLSession, didReceive
+                    challenge: URLAuthenticationChallenge,
+                    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
+        if let serverTrust = challenge.protectionSpace.serverTrust {
+            let certificatesHandler = CertificatesHandler(serverTrust: serverTrust)
+            certificatesHandler.handleCertificates { certificates in
```

Comment:
It's a bit misleading to need a completion handler for `handleCertificates` since no async work is being done. Maybe this method should simply return `[Certificate]`, which can then be passed to the `CertificateDelegate`'s `completion`.

That way it's 100% clear from looking at this method that all code paths definitely call's the delegate's `completion`.